src/time_integrator.f90

```diff
@@ -55,23 +63,175 @@ end function constructor
   subroutine step(self, u, v, w, du, dv, dw, dt)
     implicit none
 
-    class(time_intg_t), intent(in) :: self
-    class(field_t), intent(inout) :: u, v, w
-    class(field_t), intent(in) :: du, dv, dw
+    class(time_intg_t), intent(inout) :: self
+    class(field_t), target, intent(inout) :: u, v, w
+    class(field_t), target, intent(in) :: du, dv, dw
 
     real(dp), intent(in) :: dt
 
-    call self%backend%vecadd(dt, du, 1._dp, u)
-    call self%backend%vecadd(dt, dv, 1._dp, v)
-    call self%backend%vecadd(dt, dw, 1._dp, w)
+    ! assign pointer to variables
+    self%curr(1)%ptr => u
+    self%curr(2)%ptr => v
+    self%curr(3)%ptr => w
+
+    ! assign pointer to variables
+    self%deriv(1)%ptr => du
+    self%deriv(2)%ptr => dv
+    self%deriv(3)%ptr => dw
+
+    ! First-order
```

Comment:
This feels quite verbose and repetitive, can we not use something like:
```Fortran
order = min(self%istep +1, self%order)
select case (order)
  case (1)
    call self%adams_bashforth_1st(dt)
  case (2)  
    call self%adams_bashforth_2nd(dt)
    ...
  
```

Or even have the order be an argument of the `self%adams_bashforth()` function and have the logic there.